pkg/controller/controllerutil/controllerutil.go

```diff
@@ -46,3 +50,56 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
 	object.SetOwnerReferences(append(object.GetOwnerReferences(), ref))
 	return nil
 }
+
+// OperationType is the action result of a CreateOrUpdate call
+type OperationType string
+
+const ( // They should complete the sentence "Deployment default/foo has been ..."
+	// OperationNoop means that the resource has not been changed
+	OperationNoop = "unchanged"
+	// OperationCreated means that a new resource has been created
+	OperationCreated = "created"
+	// OperationUpdated means that an existing resource has been updated
+	OperationUpdated = "updated"
+)
+
+// CreateOrUpdate creates or updates a kuberenes resource. It takes in a key and
+// a placeholder for the existing object and returns the modified object
+func CreateOrUpdate(ctx context.Context, c client.Client, key client.ObjectKey, existing runtime.Object, t TransformFn) (runtime.Object, OperationType, error) {
```

Comment:
our client methods tend not to return objects, and instead operate "in-place" on them.  This should do so as well